File: src/db/repo.ts
import { db } from './db';
import type { Category, CityMap, Transaction, Trip } from './schema';

// Default categories created for every new trip (categories are per-trip).
export const DEFAULT_CATEGORIES: Omit<Category, 'id' | 'tripId'>[] = [
  { name: 'Hospedagem', color: '#0E8C6B', icon: 'bed', sortOrder: 0 },
  { name: 'Passagem', color: '#B8860B', icon: 'plane', sortOrder: 1 },
  { name: 'Transporte', color: '#C2540D', icon: 'car', sortOrder: 2 },
  { name: 'Alimentação', color: '#C1352E', icon: 'food', sortOrder: 3 },
  { name: 'Compras', color: '#B23368', icon: 'shopping', sortOrder: 4 },
  { name: 'Brindes', color: '#7D1F44', icon: 'gift', sortOrder: 5 },
  { name: 'Turismo', color: '#8A7220', icon: 'ticket', sortOrder: 6 },
  { name: 'Genéricos de viagem', color: '#7A4A2A', icon: 'luggage', sortOrder: 7 },
  { name: 'Cannabis', color: '#3D8B4C', icon: 'leaf', sortOrder: 8 },
  { name: 'Outros', color: '#5C5650', icon: 'bookmark', sortOrder: 9 },
];

// ---- trips -------------------------------------------------------------------
export async function createTrip(
  data: Omit<Trip, 'id' | 'createdAt' | 'currency'> & { currency?: string },
): Promise<string> {
  const id = `trip_${crypto.randomUUID()}`;
  await db.trips.add({
    ...data,
    id,
    currency: data.currency ?? 'BRL',
    cities: data.cities ?? {},
    createdAt: new Date().toISOString(),
  });
  // seed this trip with its own default categories
  await db.categories.bulkAdd(
    DEFAULT_CATEGORIES.map((c) => ({ ...c, id: `cat_${crypto.randomUUID()}`, tripId: id })),
  );
  return id;
}

export const updateTrip = (id: string, patch: Partial<Trip>) => db.trips.update(id, patch);

// Permanently delete a trip and everything that belongs to it.
export async function deleteTrip(id: string) {
  await db.transaction('rw', [db.trips, db.transactions, db.categories, db.rules], async () => {
    const catIds = (await db.categories.where('tripId').equals(id).primaryKeys()) as string[];
    if (catIds.length) await db.rules.where('categoryId').anyOf(catIds).delete();
    await db.transactions.where('tripId').equals(id).delete();
    await db.categories.where('tripId').equals(id).delete();
    await db.trips.delete(id);
  });
}

// Set the city of a day. Stored on the trip (a day = a city), so it persists
// even for days without transactions and even when cleared to empty.
export async function setTripCity(tripId: string, date: string, city: string) {
  const trip = await db.trips.get(tripId);
  if (!trip) return;
  const cities: CityMap = { ...(trip.cities ?? {}) };
  const v = city.trim();
  if (v) cities[date] = v;
  else delete cities[date];
  await db.trips.update(tripId, { cities });
}

// ---- transactions ------------------------------------------------------------
export async function addTransaction(
  t: Omit<Transaction, 'id' | 'createdAt'>,
): Promise<string> {
  const id = `tx_${crypto.randomUUID()}`;
  await db.transactions.add({ ...t, id, createdAt: new Date().toISOString() });
  return id;
}

export const updateTransaction = (id: string, patch: Partial<Transaction>) =>
  db.transactions.update(id, patch);

export const deleteTransaction = (id: string) => db.transactions.delete(id);

export const deleteTransactions = (ids: string[]) => db.transactions.bulkDelete(ids);

// Inserts many transactions at once (statement import) — same shape as
// addTransaction, batched into a single Dexie call.
export async function bulkAddTransactions(
  rows: Omit<Transaction, 'id' | 'createdAt'>[],
): Promise<string[]> {
  const createdAt = new Date().toISOString();
  const withIds = rows.map((r) => ({ ...r, id: `tx_${crypto.randomUUID()}`, createdAt }));
  await db.transactions.bulkAdd(withIds);
  return withIds.map((r) => r.id);
}

// ---- categories --------------------------------------------------------------
export async function addCategory(
  data: Omit<Category, 'id' | 'sortOrder'> & { sortOrder?: number },
): Promise<string> {
  const id = `cat_${crypto.randomUUID()}`;
  const existing = await db.categories.where('tripId').equals(data.tripId).toArray();
  const maxOrder = existing.reduce((m, c) => Math.max(m, c.sortOrder), -1);
  await db.categories.add({ ...data, id, sortOrder: data.sortOrder ?? maxOrder + 1 });
  return id;
}

export const updateCategory = (id: string, patch: Partial<Category>) =>
  db.categories.update(id, patch);

// Deleting a category clears it from any transaction that used it.
export async function deleteCategory(id: string) {
  await db.transaction('rw', [db.categories, db.transactions, db.rules], async () => {
    await db.transactions.where('categoryId').equals(id).modify({ categoryId: null });
    await db.rules.where('categoryId').equals(id).delete();
    await db.categories.delete(id);
  });
}

File: flutter_app/lib/data/repo.dart
import 'dart:convert';
import 'package:drift/drift.dart';
import 'package:centavoo/data/database.dart';
import 'package:centavoo/models/transaction.dart' show periodBefore, periodDuring;

class DefaultCategory {
  final String name;
  final String color;
  final String icon;
  const DefaultCategory(this.name, this.color, this.icon);
}

const defaultCategories = [
  DefaultCategory('Hospedagem', '#0E8C6B', 'bed'),
  DefaultCategory('Passagem', '#B8860B', 'plane'),
  DefaultCategory('Transporte', '#C2540D', 'car'),
  DefaultCategory('Alimentação', '#C1352E', 'food'),
  DefaultCategory('Compras', '#B23368', 'shopping'),
  DefaultCategory('Brindes', '#7D1F44', 'gift'),
  DefaultCategory('Turismo', '#8A7220', 'ticket'),
  DefaultCategory('Genéricos de viagem', '#7A4A2A', 'luggage'),
  DefaultCategory('Outros', '#5C5650', 'bookmark'),
];

String _newId(String prefix) => '${prefix}_${DateTime.now().microsecondsSinceEpoch}_${identityHashCode(Object())}';

Future<String> createTrip(
  AppDatabase db, {
  required String name,
  String? destination,
  String? startDate,
  String? endDate,
  String currency = 'BRL',
  String? notes,
}) async {
  final id = _newId('trip');
  await db.into(db.tripsTable).insert(TripsTableCompanion.insert(
        id: id,
        name: name,
        destination: Value(destination),
        startDate: Value(startDate),
        endDate: Value(endDate),
        currency: currency,
        notes: Value(notes),
        createdAt: DateTime.now().toIso8601String(),
      ));
  await db.batch((batch) {
    batch.insertAll(db.categoriesTable, [
      for (var i = 0; i < defaultCategories.length; i++)
        CategoriesTableCompanion.insert(
          id: _newId('cat_$i'),
          tripId: id,
          name: defaultCategories[i].name,
          color: defaultCategories[i].color,
          icon: Value(defaultCategories[i].icon),
          sortOrder: i,
        ),
    ]);
  });
  return id;
}

Future<void> deleteTrip(AppDatabase db, String id) async {
  await db.transaction(() async {
    final catIds = await (db.selectOnly(db.categoriesTable)
          ..addColumns([db.categoriesTable.id])
          ..where(db.categoriesTable.tripId.equals(id)))
        .map((row) => row.read(db.categoriesTable.id)!)
        .get();
    if (catIds.isNotEmpty) {
      await (db.delete(db.categoryRulesTable)..where((r) => r.categoryId.isIn(catIds))).go();
    }
    await (db.delete(db.transactionsTable)..where((t) => t.tripId.equals(id))).go();
    await (db.delete(db.categoriesTable)..where((c) => c.tripId.equals(id))).go();
    await (db.delete(db.tripsTable)..where((t) => t.id.equals(id))).go();
  });
}

Future<void> setTripCityRange(AppDatabase db, String tripId, List<String> days, String city) async {
  final trip = await (db.select(db.tripsTable)..where((t) => t.id.equals(tripId))).getSingleOrNull();
  if (trip == null) return;
  final cities = Map<String, String>.from(jsonDecode(trip.citiesJson) as Map);
  final v = city.trim();
  for (final d in days) {
    if (v.isNotEmpty) {
      cities[d] = v;
    } else {
      cities.remove(d);
    }
  }
  await (db.update(db.tripsTable)..where((t) => t.id.equals(tripId)))
      .write(TripsTableCompanion(citiesJson: Value(jsonEncode(cities))));
}

Future<void> updateTripDetails(
  AppDatabase db,
  String tripId, {
  required String name,
  String? destination,
  String? startDate,
  String? endDate,
}) async {
  await (db.update(db.tripsTable)..where((t) => t.id.equals(tripId))).write(
    TripsTableCompanion(
      name: Value(name),
      destination: Value(destination),
      startDate: Value(startDate),
      endDate: Value(endDate),
    ),
  );
}

Future<void> updateTransaction(AppDatabase db, String id, TransactionsTableCompanion patch) async {
  await (db.update(db.transactionsTable)..where((t) => t.id.equals(id))).write(patch);
}

Future<void> updateTripCityList(AppDatabase db, String tripId, List<String> cityList) async {
  await (db.update(db.tripsTable)..where((t) => t.id.equals(tripId)))
      .write(TripsTableCompanion(cityListJson: Value(jsonEncode(cityList))));
}

Future<String> addTransaction(AppDatabase db, TransactionsTableCompanion data) async {
  final id = _newId('tx');
  await db.into(db.transactionsTable).insert(
        data.copyWith(id: Value(id), createdAt: Value(DateTime.now().toIso8601String())),
      );
  return id;
}

Future<List<String>> bulkAddTransactions(AppDatabase db, List<TransactionsTableCompanion> rows) async {
  final createdAt = DateTime.now().toIso8601String();
  final ids = [for (var i = 0; i < rows.length; i++) _newId('tx_$i')];
  await db.batch((batch) {
    batch.insertAll(db.transactionsTable, [
      for (var i = 0; i < rows.length; i++)
        rows[i].copyWith(id: Value(ids[i]), createdAt: Value(createdAt)),
    ]);
  });
  return ids;
}

Future<void> deleteTransaction(AppDatabase db, String id) async {
  await (db.delete(db.transactionsTable)..where((t) => t.id.equals(id))).go();
}

Future<void> deleteTransactions(AppDatabase db, List<String> ids) async {
  await (db.delete(db.transactionsTable)..where((t) => t.id.isIn(ids))).go();
}

String? _periodForDate(String? date, String startDate) {
  if (date == null) return null;
  return date.compareTo(startDate) < 0 ? periodBefore : periodDuring;
}

Future<void> reassignTransactionPeriods(AppDatabase db, String tripId, String? startDate) async {
  if (startDate == null) return;
  final rows = await (db.select(db.transactionsTable)..where((t) => t.tripId.equals(tripId))).get();
  await db.transaction(() async {
    for (final row in rows) {
      final period = _periodForDate(row.date, startDate);
      if (period != null && period != row.period) {
        await (db.update(db.transactionsTable)..where((t) => t.id.equals(row.id)))
            .write(TransactionsTableCompanion(period: Value(period)));
      }
    }
  });
}

Future<String> addCategory(
  AppDatabase db, {
  required String tripId,
  required String name,
  required String color,
  String? icon,
  int? sortOrder,
}) async {
  final id = _newId('cat');
  final existing = await (db.select(db.categoriesTable)..where((c) => c.tripId.equals(tripId))).get();
  final maxOrder = existing.fold<int>(-1, (m, c) => c.sortOrder > m ? c.sortOrder : m);
  await db.into(db.categoriesTable).insert(CategoriesTableCompanion.insert(
        id: id,
        tripId: tripId,
        name: name,
        color: color,
        icon: Value(icon),
        sortOrder: sortOrder ?? maxOrder + 1,
      ));
  return id;
}

Future<void> updateCategory(AppDatabase db, String id, CategoriesTableCompanion patch) async {
  await (db.update(db.categoriesTable)..where((c) => c.id.equals(id))).write(patch);
}

Future<void> deleteCategory(AppDatabase db, String id) async {
  await db.transaction(() async {
    await (db.update(db.transactionsTable)..where((t) => t.categoryId.equals(id)))
        .write(const TransactionsTableCompanion(categoryId: Value(null)));
    await (db.delete(db.categoryRulesTable)..where((r) => r.categoryId.equals(id))).go();
    await (db.delete(db.categoriesTable)..where((c) => c.id.equals(id))).go();
  });
}
